packages/@aws-cdk/aws-rds/lib/instance.ts

```diff
@@ -897,23 +897,38 @@ abstract class DatabaseInstanceSource extends DatabaseInstanceNew implements IDa
    *                if you want to override the defaults
    */
   public addRotationSingleUser(options: RotationSingleUserOptions = {}): secretsmanager.SecretRotation {
+    if (!options.secret || options.secret === this.secret) {
+      return this.addSingleUserRotationForMasterSecret(options);
+    }
+
+    const id = 'RotationSingleUser';
+    return this.addSingleUserRotationForSecret(options.secret as unknown as Construct, id, options.secret, options);
+  }
+
+  private addSingleUserRotationForMasterSecret(options: RotationSingleUserOptions): secretsmanager.SecretRotation {
     if (!this.secret) {
-      throw new Error('Cannot add single user rotation for an instance without secret.');
+      throw new Error('Cannot add single user rotation for an instance without master secret.');
     }
 
     const id = 'RotationSingleUser';
-    const existing = this.node.tryFindChild(id);
-    if (existing) {
-      throw new Error('A single user rotation was already added to this instance.');
+    return this.addSingleUserRotationForSecret(this, id, this.secret, options);
+  }
+
+  private addSingleUserRotationForSecret(scope: Construct, id: string, secret: secretsmanager.ISecret, options: RotationSingleUserOptions)
+    : secretsmanager.SecretRotation {
+    // TODO: this is a hack! what's the right way to do it?
+    if ((scope as any).node.tryFindChild(id)) {
+      const secretId = this === scope ? 'master' : 'additional'; // Is there a better way to do this?
+      throw new Error(`A single user rotation for ${secretId} secret is already added to this instance.`);
     }
 
-    return new secretsmanager.SecretRotation(this, id, {
-      secret: this.secret,
+    return new secretsmanager.SecretRotation(scope, id, {
       application: this.singleUserRotationApplication,
       vpc: this.vpc,
       vpcSubnets: this.vpcPlacement,
       target: this,
       ...options,
+      secret: secret,
       excludeCharacters: options.excludeCharacters ?? DEFAULT_PASSWORD_EXCLUDE_CHARS,
     });
   }
```

Comment:
Same comments here (move the private methods below the public ones, pass `secret` through `options.secret`, remove the cast of `scope: Construct` to `any`, etc.).

Reply:
Ditto.